Add TableInfo and related model classes

mziccard opened the pull request:
This is a very big PR (but most of it is javadocs and tests) that adds `TableInfo` model class and all related classes:
- `TableSchema`
- `FieldSchema`
- `CsvOptions`
- `ExternalDataConfiguration`
- `UserDefinedFunction`

Unit tests are also added.

**Remarks**
I added the `TableSchema` class to wrap schema information as it also appears elsewhere (Query results). Schema is just a list of `FieldSchema` objects so if you prefer we can drop the class and just use `List<FieldSchema>`. I have no strong preferences.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/CsvOptions.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * Creates an {@code ExternalDataConfiguration} object.
```
s/ExternalDataConfiguration/CsvOptions/

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/CsvOptions.java:
```diff
@@ -0,0 +1,267 @@
+     * Set whether BigQuery should accept rows that are missing trailing optional columns. If
+     * {@code true}, BigQuery treats missing trailing columns as null values. If {@code false},
+     * records with missing trailing columns are treated as bad records, and if there are too many
+     * bad records, an invalid error is returned in the job result. By default, rows with missing
```
As for "too many bad records" it would be nice to have a link to `ExternalDataConfiguration.maxBadRecords`

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/CsvOptions.java:
```diff
@@ -0,0 +1,267 @@
+     * default value is UTF-8. BigQuery decodes the data after the raw, binary data has been split
+     * using the values set in {@link #quote(String)} and {@link #fieldDelimiter(String)}.
+     */
+    public Builder encoding(String encoding) {
```
Fine as is. Alternative (or addition) could be to accept a `Charset` and then use its `name()`

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/ExternalDataConfiguration.java:
```diff
@@ -0,0 +1,419 @@
+ * @see <a href="https://cloud.google.com/bigquery/federated-data-sources">Federated Data Sources
+ *     </a>
+ */
+public class ExternalDataConfiguration implements Serializable {
```
Would you have something similar for a data configuration of a load job?

mziccard replied:
Yes, a load job has its own configuration (which is slightly different). This class is also used to define [external data sources to a query job](https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.query.tableDefinitions), which are query-scoped federated tables (described [here](https://cloud.google.com/bigquery/federated-data-sources#cli))

aozarov commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/UserDefinedFunctionTest.java:
```diff
@@ -0,0 +1,7 @@
+ * Created by mziccardi on 11/30/15.
+ */
+public class UserDefinedFunctionTest {
+}
```
implement or add TODO or remove.

mziccard replied:
Ooops, must have forgot to re-add after implementing. Weird. Will fix.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/ExternalDataConfiguration.java:
```diff
@@ -0,0 +1,419 @@
+    }
+
+    /**
+     * Sets the source format of the external data.
```
Maybe we could copy some of the text from the link?
BTW, https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.load.sourceFormat seems to include more options...

mziccard replied:
Yes I noticed that, and also an extract job supports [different destination formats](https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.extract.destinationFormat).

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/ExternalDataConfiguration.java:
```diff
@@ -0,0 +1,419 @@
+     * Sets additional properties to be used to parse CSV data (used when
+     * {@link #sourceFormat(String)} is set to CSV).
+     */
+    public Builder csvOptions(CsvOptions csvOptions) {
```
What is your plan to support future format specific options?

mziccard replied:
The approach I adopted so far for *pick-one* fields is to have all fields in the top level class and design the builder and factory methods so that only one is initialized at a time. It would be something like:
```java
public static ExternalDataConfiguration of(String sourceUri, TableSchema schema, CsvOptions options) {
  return builder(sourceUri, schema, options).format("CSV").build();
}

public static ExternalDataConfiguration of(String sourceUri, TableSchema schema, JsonOptions options) {
  return builder(sourceUri, schema, options).format("NEWLINE_DELEMITED_JSON").build();
}
```

This should enforce a "good" construction.

aozarov replied:
Yes, that helps but one could also have access and use the `Builder` directly.

Did you consider a base class for `FormatOptions` with `type()` returns enum or string of possible types? then one need to set the formatOptions (probably instead of supplying `sourceFormat`) and you can use the enum in the toPb...?

mziccard replied:
I thought about that but in general I don't like requiring the user to do casting to access some data:
```java
FormatOptions options = externalDataConfiguration.formatOptions();
// to access real options:
CsvOptions csvOptions = (CsvOptions) options;
```
However, if you think having `FormatOptions` is better I am happy to change.

mziccard replied:
In the end I gave this a try in 79001cf046ca1f33427231655dd457cd41015e09, what do you think?

aozarov replied:
I like it! I added few further suggestions.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldSchema.java:
```diff
@@ -0,0 +1,287 @@
+import java.util.Objects;
+
+/**
+ * Google Bigquery schema for a Table field. Several data types and modes are supported.
```
I find this javadoc a bit confusing. what does it mean "Table field". 
What is the purpose of the "Several data types and modes are supported" description and if that is only a subset where can I get the complete list.

aozarov replied:
Ok, I don't find the first part ("Table field") as confusing now, but few suggestions:
(1) we should be consistent with name style Bigquery -> BigQuery
(2) Maybe we should say something like "Google BigQuery field schema" (schema or description) ", which is part of a TableSchema".
(3) did you consider these names TableSchema -> Table, FieldSchema -> Field  or do you think that may collide with the values options? Another possibility is Schema (for TableSchema) and Field or FieldSchema.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldSchema.java:
```diff
@@ -0,0 +1,287 @@
+     * Sets the mode of the field. By default {@link Mode#NULLABLE} is used.
+     */
+    public Builder mode(Mode mode) {
+      this.mode = mode;
```
Should we consider null values for patch operation or is that not supported?

mziccard replied:
The only [supported changes](https://cloud.google.com/bigquery/docs/tables#updateschema) to the schema of an existing table are:
- Adding NULLABLE or REPEATED fields at the end of the schema
- Making REQUIRED fields NULLABLE

Even though "unsetting" a REQUIRED mode should set it to the default (NULLABLE) I think requiring explicit values here is better.

aozarov replied:
Honestly, I think if we are going to use patch for updating a [`Table`](https://github.com/GoogleCloudPlatform/gcloud-java/blob/bigquery/gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigQueryRpc.java#L153) we should probably
not relax the implementation based on current restriction as that might get relaxed in the future
and break the code.

If you agree, we don't have to do it now but rather create an issue for it.

mziccard replied:
Why do you think updates will break the code? Here we are simply not allowing a user to unset a field's mode, which does not make sense at the moment. Am I missing something?

aozarov replied:
My comment was more generic rather than specific. I agree that it may not be that useful
to support reseting a mode to default (NULLABLE) but if Builder allows setting `null` why not
respect it in patch operations? Same applies to description...

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldSchema.java:
```diff
@@ -0,0 +1,287 @@
+     * Creates an {@code FieldSchema} object.
+     */
+    public FieldSchema build() {
+      FieldSchema fieldSchema = new FieldSchema(this);
```
Alternative: 

```java
checkNotNull(fieldSchema.name);
checkNotNull(fieldSchema.type);
return new FieldSchema(this);
```

Also, should you check that fields is only populated when type is RECORD?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldSchema.java:
```diff
@@ -0,0 +1,287 @@
+   * @see <a href="https://cloud.google.com/bigquery/preparing-data-for-bigquery#datatypes">
+   *     Data Types</a>
+   */
+  public enum Type {
```
Another option that works nicer with RECORDS (and fields) is to make that a class with private constructor.
The class could have an an inner enum `Value` with the same set  of values.
Class has a factory method for each type `string()`, `integer()`, ... record(FieldSchema...)`.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldSchema.java:
```diff
@@ -0,0 +1,287 @@
+     * Sets the nested schema fields if {@link #type(Type)} is set to {@link Type#RECORD}.
+     */
+    Builder fields(Iterable<FieldSchema> fields) {
+      this.fields = fields != null ? ImmutableList.copyOf(fields) : null;
```
Is it valid to have a record with no fields?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableSchema.java:
```diff
@@ -0,0 +1,151 @@
+   * Returns a builder for the {@code TableSchema} object.
+   */
+  public Builder toBuilder() {
+    Builder builder = new Builder();
```
Alternative:
```java
return builder().fields(fields);
```

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/UserDefinedFunction.java:
```diff
@@ -0,0 +1,158 @@
+   * Creates a Google Cloud BigQuery user-defined function given a code blob.
+   */
+  public static UserDefinedFunction inline(String functionDefinition) {
+    return new InlineFunction(functionDefinition);
```
should we make the specific constructors package scope?

aozarov replied:
Also, in its current state I am not sure there is even a value in exposing the specific sub-classes 'UriFunction` and `UserDefinedFunction`.

mziccard commented:
Fixed the commends, main changes are:
- `TableSchema` renamed to `Schema` and `FieldSchema` renamed to `Field`
- Changed `Type` enum to class in `Field` with factory methods for all types
- Add setter and builder method that take fields as a vararg for `Schema`
- Fixed all javadocs comments and updated tests according to changes

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Field.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    /**
+     * Creates an {@code Field} object.
```
s/ an / a / ?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Field.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    /**
+     * Sets the mode of the field. By default {@link Mode#NULLABLE} is used.
```
Fine as is (and I don't remember what we typically say, so preference is for consistency) but another way to say it would be "Sets the mode of the field. When not specified {@lilnk Mode#Nullable} will be used."

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Field.java:
```diff
@@ -0,0 +1,370 @@
+    return description;
+  }
+
+  /**
```
Though could be considered as a "convenient" method my preference would be not to have it.
Feel free to leave as is if you think otherwise.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Schema.java:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  public static Schema of(Iterable<Field> fields) {
+    checkNotNull(fields);
```
will be checked in `Builder.fields` setter.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/UserDefinedFunction.java:
```diff
@@ -0,0 +1,151 @@
+   * a code blob. If {@link #type()} is {@link Type#FROM_URI} this method returns a Google Cloud
+   * Storage URI (e.g. gs://bucket/path).
+   */
+  public String functionDefinition() {
```
How about `content` considering the scope `UserDefinedFunction` and the fact that it may actually not be a function definition but rather a reference.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/UserDefinedFunction.java:
```diff
@@ -0,0 +1,151 @@
+  /**
+   * Returns function's definition. If {@link #type()} is {@link Type#INLINE} this method returns
+   * a code blob. If {@link #type()} is {@link Type#FROM_URI} this method returns a Google Cloud
+   * Storage URI (e.g. gs://bucket/path).
```
Is externalDataConfiguration.sourceUris also using the same format "gs://bucket/path" ?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableInfo.java:
```diff
@@ -0,0 +1,741 @@
+  /**
+   * The table type.
+   */
+  public enum Type {
```
Did you consider using sub-classes?

mziccard replied:
Do you think there's a need for this?

aozarov replied:
I feel that using a standard Java inheritance would be a nicer user experience than having a flat type with all possible setter/getter (merged from all types TABLE, EXTERNAL and VIEW).

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableInfo.java:
```diff
@@ -0,0 +1,741 @@
+    }
+
+    /**
+     * Sets a user-friendly name for the dataset.
```
s/dataset/table/

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableInfo.java:
```diff
@@ -0,0 +1,741 @@
+    }
+
+    /**
+     * Sets the table's schema. Providing a schema is not necessary when {@link #viewQuery} is
```
I assume that this is also not necessary for EXTERNAL (when ExternalDataConfiguration is provided)?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableInfo.java:
```diff
@@ -0,0 +1,741 @@
+   * Returns a user-friendly name for the table.
+   */
+  public String friendlyName() {
+    return friendlyName;
```
``` return Data.isNull(friendllyName) ? null : friendlyName;``` here and in other places where
we use `Data.nullOf`

aozarov replied:
I think this is still not fixed.

aozarov commented:
Looks good. Few more comments and suggestions.

aozarov commented:
See my comment regarding `CsvOptions` (I like the change you made and suggested to take it slightly further), the possible bug in `TableInfo` when nulls are passed in the builder and later retrieved by the built instance and also my suggestion to use inheritance for TableInfo as well (if not inheritence maybe we can consider something similar to what you did with `FormatOptions`).

Also, I feel like an open comment regarding `patch` and `ExternalDataConfiguration` got lost somehow.

mziccard commented:
I see where you are going with the hierarchies. The fact is that i really don't like that the user has to cast values to get to real data. It might be me but I really consider casting an unnecessary pain in the neck. The main point is that if we have:
```java
@SuppressWarnings("unchecked")
public <F extends FormatOptions> F formatOptions() {
  return (F) formatOptions;
}
```
To use this method a user must not only read its javadoc but he needs also to be aware of the `FormatOptions` hierarchy. If he does't know the expected type, in fact, he has to do type-checking to figure out how to use the return type. Avoiding type-checking and casting is the reason why I provided both `format` and `csvOptions` getters.

The same reasoning (even worse maybe) applies to `TableInfo` type. We might want to have something like:
```java
public abstract class TableType {

  static class Table extends TableType {
    // ...
  }
  static class View {
    // ...
  }
  static class External {
    // ...
  }
  public static Table of(Schema schema);
  public static View of (String query, List<UserDefinedFunction> functions);
  public static External of(ExternalDataConfiguration configuration);
}

public class TableInfo {
  // ...
  public static Builder builder(TableType type);
  public static ViewInfo of(TableType type);
}
```
This is ok but if you ask me I *would not* remove all `schema()`, `query()`, `userDefinedFunctions()` and `externalConfiguration()` getters from `TableInfo` if favor of `public <T extends ViewType>type()` for the same reasoning I did above for `CsvOptions`. Even worse, here the information wrapped in types is not disjoint, the table's schema in fact is returned by the BigQuery service for all 3 types of tables.

To sum things up: I am in favor of using hierarchies to make it easier for the user to build an object right, I am not in favor of having getters that require type-checking. I am open for discussion :)

> Also, I feel like an open comment regarding patch and ExternalDataConfiguration got lost somehow.

It got lost in fact, my answer is [here](https://github.com/GoogleCloudPlatform/gcloud-java/pull/423#discussion_r46421030).

aozarov commented:
I think javadoc, which shoes class hierarchies, and making the set of format type constants more public (maybe not enum, but there are other ways can do it) could help with knowing which one should be used.

Also, I would expect the user to know it as (similar to the schema) it probably set it.
I only mentioned that method as a "convenient" getter that avoid explicit casting (which
we also do in other places).

>Avoiding type-checking and casting is the reason why I provided both format and csvOptions getters.

If someone was calling `csvOptions` it is because it already knew it was a `csvOptions`. Yes,
one could call all xxxOptions and check for not `null` but this does not look better to me than
`instanceof`.

If we assume that more and more formats are going to be supported (which, I think, is the reason we don't use enum for `type`), than in that case it would mean adding more and more `xxxOptions` get methods on `ExternalDataConfiguration`. Is that desired?

As for `TableInfo` I think here we should not expect proliferation of types (if any), so
having the, currently 3, specific getters is not a problem (but I don't see that as an improvement) however flattening *all* options (such as `userDefinedFunctions`) in `TableInfo` does not seem
right to me (and force the user to know what belongs to what).  In my mind it is prefered to have common stuff in common class and specific stuff in its specific place. BTW, did you consider to make `TableInfo` abstract and have 3 concrete implementation for it?

mziccard commented:
I obey :) Kidding, I don't completely agree but could discuss this for days and there's no point in stalling this any longer.
I hope I caught all comments, let me recap:
- f84a3b1: Fixes the [comment](https://github.com/GoogleCloudPlatform/gcloud-java/pull/423#discussion_r46231753) on unsetting `Field` mode and description
- bd58ee5 Removes [redundant getters](https://github.com/GoogleCloudPlatform/gcloud-java/pull/423#commitcomment-14750946) from ExternalDataConfiguration. Makes `FormatOptions.csv` return a `CsvOptions` object (comment [here](https://github.com/GoogleCloudPlatform/gcloud-java/pull/423#commitcomment-14750781))
- 66972df: Adds the `TableType` class with 2 subclasses `View` and `ExternalTable`. This class is used in `TableInfo` [to wrap table's fields that depends on the type](https://github.com/GoogleCloudPlatform/gcloud-java/pull/423#discussion_r46366010).

aozarov commented:
I like this one better! (but still have few suggestions).

Next time when you do not completely agree I suggest you create an issue (marked as question) and we can continue to discuss it there (and solicit more opinions) while keeping the code as is.

mziccard commented:
Some more changes:
- ae2a484: Removes TableType and creates a hierarchy of `TableInfo`. If you are wondering why `BaseTableInfo` is not abstract it's because `listTables` return partial information on tables that does not allow to say whether they are normal tables, views or external tables. So I thought to use `BaseTableInfo` for them
- 338ede1: Fixes the `mode` issue and another minor comment

aozarov commented:
I really like it!! Few nits.
